src/keys/binary_key.rs

```diff
@@ -0,0 +1,277 @@
+use super::Key;
+
+/// Represents a key whose basic unit is the bit. This is meant to be
+/// used at the key in binary trees.
+///
+/// Bits are stored inside an array of bytes, and are read left to
+/// right. The structure also contains the start and end offset of
+/// the bit field.
+///
+/// # Example structure
+///
+/// The following code will create a bit field with 6 bits spread
+/// over two integers.
+///
+/// ```
+/// use multiproof_rs::keys::BinaryKey;
+/// let bitkey = BinaryKey(vec![0x55; 2], 3, 6);
+/// ```
+///
+/// The internal representation is therefore:
+///
+/// ```text
+/// byte #    |     1        2
+/// bit  #    | 76543210 76543210
+/// ----------+------------------
+/// bit value | 01010101 01010101
+/// offsets   |     ^     ^
+///           |   start  end
+/// ```
+#[derive(Debug, PartialEq, Clone)]
+pub struct BinaryKey(pub Vec<u8>, pub usize, pub usize); // (key data, offset in first byte, offset in last byte)
```

Comment:
Suggestion: instead of exposing the underlying byte-based structure, let the calling code use general bit offsets. By that I mean the first bit from left has offset 0 and if the key length is 16, the last bit (rightmost) has offset 15. This feels more intuitive for me.
Although I can see when you repeatedly want to slice the key, then those offsets change.